### ISSUE TITLE
com.expressvpn.vpn

### ISSUE THREAD

**Issue by ilyesgasmi** [httptoolkit/frida-android-unpinning#37]
No luck using the script on the expressvpn app. I tried reverse engineering the app and writing a frida script myself, but with no success. I wasn't able to figure out the functions needed to bypass the SSL verification. I would appreciate any help. Thank you

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/64224194-d555-402f-ac37-8b051677df8f)

**Comment by pimterry** [comment]
Have you see https://httptoolkit.com/blog/android-reverse-engineering/? If the script doesn't work, the only option to intercept your traffic is to reverse engineer the app I'm afraid.

If this doesn't work, do please share any details you find along the way for each of the steps there (error logs for this failure, notable references to relevant system types, any other ADB output, results for the suggested strings to search for, etc). It's almost always possible to disable pinning, and the steps there should at least guide you to the relevant code. Do share anything you come up to here, once you manage to find the relevant code I'm happy to help put together a Frida hook to disable it.

**Comment by ilyesgasmi** [comment]
Yes, I have seen the blog post. It was very insightful and it is what lead me to try myself, but as I said earlier I haven't been able to make a functional script (or any script at all). Here are my attempts:
Using a rooted memu emulator and connecting to frida using adb. I followed the blog post for all configuration.

I decompiled the app using jadx cli tool. I ended up with a sources and a ressources folder, unfortunately all with random package names: 

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/2f47a8ce-a68a-4c27-b5b1-e461c6dff682)

Opening the files with vscode and using the search function, I found that there were many references to the certificate pinner: 

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/d0be740e-de30-4056-8408-203bf4c6c0ae)

Also, browsing through the code I found an interesting method that might have been useful: 

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/eb49e741-9df1-4dc0-8395-3ca53d3e6bef)

There I encountered my first issue: My class was not being recognized:

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/bd6ce753-ef7a-4f31-b7ef-6e073292ec37)

I then used this script to view all the classes that were loaded

``
Java.perform(function() {
  var classes = Java.enumerateLoadedClassesSync();
  classes.forEach(function(className) {
    console.log(className);
  });
});
``

And I was able to find my class! 

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/4379b12e-318d-40cd-bd3d-ef9da66308bb)

After that, I tried again to run the script, and to my surprise the output changed. I got this output: 

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/934a6cea-0af8-4590-b5d3-1a1f94d00296)

I have no idea if that means the script is properly hooked, and simply not working, if it didn't.
Thank you for your help. I greatly appreciate the help

**Comment by pimterry** [comment]
Hmm, I'm not sure what would cause the missing class error. It might be that some classes are loaded somewhat dynamically, so weren't yet loaded in the app when you first ran the command, and were available later. You may need to start the app, and then delay hooking the methods until slightly later (maybe even by making it fail, and then hooking the methods afterwards).

Overall it looks like this app has obfuscated the sources of its libraries, not just its own code, which is unusual but does explain why the hooks here aren't working. One good thing though is that you can see that OkHTTP is included (some OkHTTP strings are shown in your find results). This is commonly used for HTTP in apps themselves (not so much in libraries) and so if it's included it's fairly likely that that's the main client they're using, and that's what you need to patch.

Although the [existing OkHTTP hooks](https://github.com/httptoolkit/frida-android-unpinning/blob/4d477da8c58c353a0290ec4829a1de4ca1ca5ae5/frida-script.js#L161-L206) from this script won't work, because the class & method names have all been changed, if you can find the equivalent code in your app you should be able to use the same hook as in this script, just swapping out the class & method names. The `check` method our hooks patch originally looks like [this](https://github.com/square/okhttp/blob/8d811228662e7c9d556f7342d6410c113699e8e8/okhttp/src/jvmMain/kotlin/okhttp3/CertificatePinner.kt#L140-L180) - if you can find the equivalent in your obfuscated code, and then swap the obfuscated class & method names into the script, I think that should work.

I'd strongly recommend looking at the ADB logs too - that will often have exceptions listed when connections fail, and those may lead you directly to the right classes.

> I have no idea if that means the script is properly hooked, and simply not working, if it didn't.

To tell if your hooks are working, you can:

* Add `console.log` lines inside your hooked methods - if they're called, you'll see your log output, and that's a pretty good sign that you've hooked something useful.
* Use HTTP Toolkit (or some other intercepting proxy) and checking for failed TLS connections when the app is running. If you're seeing any then there is some kind of certificate issue (either the device doesn't trust your cert at all, or it does but there's some pinning involved).

**Comment by ilyesgasmi** [comment]
I think I found the equivalent in my code! 

![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/2c96a8e8-b42a-49d3-9ed4-d60eae8059fd)

However, I am not sure on how to translate your script to work on my class and method names. Here is my best attempt at that:

```js

// OkHTTPv3 (quadruple bypass)
try {
    // Bypass OkHTTPv3 {1}
    const okhttp3_Activity_1 = Java.use('s50.g');
    okhttp3_Activity_1.b.overload('java.lang.String', 'java.util.List').implementation = function (a, b) {
      console.log('  --> Bypassing OkHTTPv3 (list): ' + a);
      return;
    };
    console.log('[+] OkHTTPv3 (list)');
  } catch (err) {
    console.log('[ ] OkHTTPv3 (list)', err);
  }
  
  try {
    // Bypass OkHTTPv3 {2}
    // This method of CertificatePinner.check could be found in some old Android app
    const okhttp3_Activity_2 = Java.use('s50.g');
    okhttp3_Activity_2.b.overload('java.lang.String', 'java.security.cert.Certificate').implementation = function (a, b) {
      console.log('  --> Bypassing OkHTTPv3 (cert): ' + a);
      return;
    };
    console.log('[+] OkHTTPv3 (cert)');
  } catch (err) {
    console.log('[ ] OkHTTPv3 (cert)', err);
  }
  
  try {
    // Bypass OkHTTPv3 {3}
    const okhttp3_Activity_3 = Java.use('s50.g');
    okhttp3_Activity_3.b.overload('java.lang.String', '[Ljava.security.cert.Certificate;').implementation = function (a, b) {
      console.log('  --> Bypassing OkHTTPv3 (cert array): ' + a);
      return;
    };
    console.log('[+] OkHTTPv3 (cert array)');
  } catch (err) {
    console.log('[ ] OkHTTPv3 (cert array)', err);
  }
  
  try {
    // Bypass OkHTTPv3 {4}
    const okhttp3_Activity_4 = Java.use('s50.g');
    okhttp3_Activity_4['b$okhttp'].implementation = function (a, b) {
      console.log('  --> Bypassing OkHTTPv3 ($okhttp): ' + a);
      return;
    };
    console.log('[+] OkHTTPv3 ($okhttp)');
  } catch (err) {
    console.log('[ ] OkHTTPv3 ($okhttp)', err);
  }
  
```


  I get these errors returned: 
  
![image](https://github.com/httptoolkit/frida-android-unpinning/assets/77757014/a7c5067a-70fc-4294-83a7-7c357cbdf3a8)

I believe this means that the method java.lang.String is obfuscated, and I need to find the new method name and replace it in the script. Am I on the right path? And if so, do you know where I could fine the obfuscated method name? It doesn't seem to be in the "check" method from okhttp.
Thank you

**Comment by pimterry** [comment]
Nice! Yes that looks promising.

That error there is saying that the method name you have does match, but the argument types don't - you just need to change the argument types (as you're passing to `overload()` to match the error message (so replace the second argument, e.g `java.util.List`, with `b50.a` - I think `java.lang.String` is fine). In our hooks code, we cover multiple versions of the library, which is part of why there are multiple overloads - in this case since there is only one overload present (as the error message shows you) you'll only need one of those first three cases.

It does look like the `b$okhttp` property is wrong for the last case, I'm not totally sure how to fix that. For awkward cases, it's usually best to manually mess around with Frida to explore the code: connect with Frida, then run `Java.use('s50.g')`, and then see what properties and methods are available there.
